### ISSUE TITLE
Subscription not firing on some resources (User, Login) when created with MedplumClient methods

### ISSUE THREAD

**Issue by bnapora** [medplum/medplum#3979]
I've been creating Users with the MedplumClient and came across an issue with Subscriptions not firing when a User resource is created. 

When a User resource is created using `startGoogleLogin` (and possibly `startNewUse`r), the resource is created WITHOUT a `meta.project` value.  It does have a `project.reference` value.  The getSubscriptions method in subscriptions.ts (packages/server/src/workers/subscription.ts: Line 275) is only checking for subscriptions if there is a value in meta.project field.  If no value, then no subscriptions retrieved.  If User resources is "edited" in Admin app, then a meta.project value is added to the User resource, and a subscription is triggered.

I believe this is a bug, but not certain if should be resolved in the MedplumClient React component or in the Server.

If additional information is needed I'm happy to provide.

**Comment by codyebberson** [comment]
Hi @bnapora - thank you for filing this.

Yes, this is unfortunately a known issue / design issue.  

> My temporary solution is to modify (packages/server/src/auth/google.ts) and add: meta: projectId ? { project: projectId} : undefined, as Line 111

tl;dr - I think this is a good solution, and we should make this part of the official `auth/google` API.

--------

Long version:

First, for context, a `User` resource can be "server scoped" (sometimes called "global scoped") or "project scoped".

* Server scoped / Global scoped
    * The same `User` resource can be shared across multiple projects
    * There is a global unique constraint on `User.email` at the database layer
    * This is the default for new users via the "Register" flow
    * The motivation is to enable easy SSO with multiple `ProjectMembership`
    * Signing into the app with a server-scoped `User` and multiple `ProjectMembership` leads to the "Choose Profile" UI
* Project scoped
    * The `User` resource includes `User.project` (and indirectly `User.meta.project`)
    * The `User` is fully contained within the project
    * There is a unique constraint on `(User.email, User.project)`
    * This is the default for new "Patient" users and new "Practioner" users if and only if the Practitioner has "externalId"
    * The motivation behind project scoped users is to create a fully isolated white-labeled experience
    * Project scoped users are not intended to ever sign into the Medplum admin app

Now, with regards to subscriptions.

In general, subscriptions are project scoped, and therefore can only subscribe to resources within the same project.

Which leads to your issue.

Similar to your solution, one of the things that we have contemplated doing is ensuring that "project" is an input to all APIs that create a new user.  That way, you as the app author can always ensure that the resulting `User` is always project scoped. 

Another option is to enable "Server scoped subscriptions".  Basically a super-admin subscription that receives events across all projects.  Despite the inherent risks, that is a sometimes necessary capability.

**Comment by bnapora** [comment]
@codyebberson  Thanks for prompt reply.

**Meta.projectId in auth/google API** - should I submit a PR for this modification, or is it something you'd prefer to add through other updates?

**User Scope (Server vs Project)** - Your description of a "project" scoped user is what we are attempting to accomplish  (isolated white-label experience, no access to Medplum admin app.)  A couple primary considerations:

- We were planning to require Google Auth (since it is also used for authorization of GCP DICOM Store assets once user is logged in.)  
- We also are trying to create a new Project for each new User/Practitioner.  

We created a Bot that triggers on User CREATE, and then creates the Practitioner, Project and ProjectMembership resources.  The Bot successfully creates the resources, however GoogleAuth fails. I think I now understand it is because of the default "unique constraint" you described above:
> There is a unique constraint on (User.email, User.project)
This is the default for new "Patient" users and new "Practioner" users if and only if the Practitioner has "externalId"

In our use case the Practitioner only has an "externalId", we don't know the ProjectId, until they authenticate, but the current GoogleAuth  fails because no ProjectID provided.  Is it possible to adjust the unique constraint to be just the GoogleAuth "User.email"?  (Presumeably, the GoogleAuth routine would then supply all ProjectMemberships for that user?) 
- (My other solution was to make a User part of a "management project" that is consistent, so that we can look up the memberships, and then supply the Users specific projectID to the auth workflow.)

Subscriptions - in this use case the User Subscription is to facilitate a custom Project/Practitioner creation workflow.  This seems to fall on the edge of the standard project scoped User creation process.  My vote would be to enable "Server scoped subscriptions" to enable initial "generic" User creation using external auth (eg Google Auth).  It's likely that other Resource creation use cases where multiple projects are in play would benefit from Server scoped subscriptions.
